src/irr/asset/CImageWriterPNG.cpp

```diff
@@ -103,30 +149,45 @@ bool CImageWriterPNG::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
 		png_destroy_write_struct(&png_ptr, &info_ptr);
 		return false;
 	}
+
+	core::smart_refctd_ptr<ICPUImage> convertedImage;
+	{
+		const auto channelCount = asset::getFormatChannelCount(image->getCreationParameters().format);
+		if (channelCount == 1)
+			convertedImage = getPNGConvertedOutput<asset::EF_R8_SRGB>(image);
+		else if(channelCount == 2 || channelCount == 3)
+			convertedImage = getPNGConvertedOutput<asset::EF_R8G8B8_SRGB>(image);
+		else
+			convertedImage = getPNGConvertedOutput<asset::EF_R8G8B8A8_SRGB>(image);
```

Comment:
wondered if you could have done the `getPNGConvertedOutput` straight from the input image view, and not call `getTopImageDataForCommonWriting`

Reply:
I think we need such a function in various writers

Reply:
there will be many code duplication